File: backend/app/api/schema.py
from __future__ import annotations

from datetime import datetime
from typing import Any

from fastapi import APIRouter, HTTPException, Request
from pydantic import BaseModel
from sqlmodel import Session, select

from app.influx.schema_discovery import list_cached
from app.state.models import MeasurementAnnotation, SchemaCache

router = APIRouter(prefix="/api/schema", tags=["schema"])


class MeasurementOut(BaseModel):
    measurement: str
    description: str
    tag_keys: list[str]
    field_keys: list[dict[str, Any]]
    tag_values: dict[str, list[str]]
    updated_at: datetime


class SchemaSummary(BaseModel):
    measurements: list[MeasurementOut]
    last_refresh: datetime | None
    last_error: str | None
    is_refreshing: bool


class DescriptionUpdate(BaseModel):
    description: str


@router.get("", response_model=SchemaSummary)
async def get_schema(request: Request) -> SchemaSummary:
    engine = request.app.state.db_engine
    refresher = request.app.state.refresher
    with Session(engine) as session:
        cached = list_cached(session)
        annotations = {
            a.measurement: a.description for a in session.exec(select(MeasurementAnnotation)).all()
        }
        measurements = [
            MeasurementOut(
                measurement=row.measurement,
                description=annotations.get(row.measurement, ""),
                tag_keys=row.tag_keys,
                field_keys=row.field_keys,
                tag_values=row.tag_values,
                updated_at=row.updated_at,
            )
            for row in cached
        ]
    return SchemaSummary(
        measurements=measurements,
        last_refresh=refresher.last_run,
        last_error=refresher.last_error,
        is_refreshing=refresher.is_running,
    )


@router.put("/{measurement}/description", response_model=MeasurementOut)
async def update_description(
    measurement: str, body: DescriptionUpdate, request: Request
) -> MeasurementOut:
    engine = request.app.state.db_engine
    with Session(engine) as session:
        cached = session.get(SchemaCache, measurement)
        if cached is None:
            raise HTTPException(status_code=404, detail=f"unknown measurement: {measurement}")
        ann = session.get(MeasurementAnnotation, measurement)
        if ann is None:
            ann = MeasurementAnnotation(measurement=measurement, description=body.description)
            session.add(ann)
        else:
            ann.description = body.description
            ann.updated_at = datetime.utcnow()
            session.add(ann)
        session.commit()
        session.refresh(ann)
        return MeasurementOut(
            measurement=cached.measurement,
            description=ann.description,
            tag_keys=cached.tag_keys,
            field_keys=cached.field_keys,
            tag_values=cached.tag_values,
            updated_at=cached.updated_at,
        )


@router.post("/refresh")
async def refresh(request: Request) -> dict[str, str]:
    refresher = request.app.state.refresher
    if refresher.is_running:
        return {"status": "already-running"}
    # Fire-and-forget so the HTTP call returns quickly.
    import asyncio

    asyncio.create_task(refresher.trigger_now())
    return {"status": "started"}

File: frontend/src/api/schema.ts
import { apiUrl } from "../base";
export type FieldKey = { name: string; type: string };

export type Measurement = {
  measurement: string;
  description: string;
  tag_keys: string[];
  field_keys: FieldKey[];
  tag_values: Record<string, string[]>;
  updated_at: string;
};

export type SchemaSummary = {
  measurements: Measurement[];
  last_refresh: string | null;
  last_error: string | null;
  is_refreshing: boolean;
};

export async function fetchSchema(): Promise<SchemaSummary> {
  const res = await fetch(apiUrl("/api/schema"));
  if (!res.ok) throw new Error(`fetchSchema failed: ${res.status}`);
  return res.json();
}

export async function updateDescription(
  measurement: string,
  description: string,
): Promise<Measurement> {
  const res = await fetch(apiUrl(`/api/schema/${encodeURIComponent(measurement)}/description`), {
    method: "PUT",
    headers: { "content-type": "application/json" },
    body: JSON.stringify({ description }),
  });
  if (!res.ok) throw new Error(`updateDescription failed: ${res.status}`);
  return res.json();
}

export async function triggerRefresh(): Promise<{ status: string }> {
  const res = await fetch(apiUrl("/api/schema/refresh"), { method: "POST" });
  if (!res.ok) throw new Error(`triggerRefresh failed: ${res.status}`);
  return res.json();
}
